### PR TITLE
make-easy-loop-on-array

### PR DESCRIPTION
**Pull request by jamet-julien**
## TL;DR;

 Make a elegant loop on array 
## Username

Julien jamet: [github](https://github.com/jamet-julien)

### CODE REVIEW

**Comment by zenopopovici** [comment]
@jamet-julien: Please use english.

**Comment by jamet-julien** [comment]
Is it better ?

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+title: Make easy loop on array
+tip-number: xx
+tip-username: jamet-julien
+tip-username-profile: https://github.com/jamet-julien
+tip-tldr: Make a elegant loop on array 
+
+categories:
+    - fr
+---
+We often need to build loop, all the conditions can be huge:
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: We often need to build a loop, else the condition list would be huge.

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+change( true); //D
+change( true); //C
+change( true); //B
+change( true); //A
+change( true); //E LOOP !
+
+
+```
+
+Use operator  ```%``` can change your life and it's prettier:
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: Using the `%` operator is prettier:

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+change(); //A
+change(); //B
+change(); //C
+change(); //D
+change(); //E
+change(); //A  LOOP !
+
+change( true); //A
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: Remove space.

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: Remove space

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
+---
+layout: post
+
+title: Make easy loop on array
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: Create an easy loop using an array

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+tip-number: xx
+tip-username: jamet-julien
+tip-username-profile: https://github.com/jamet-julien
+tip-tldr: Make a elegant loop on array 
+
+categories:
+    - fr
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: en

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
+---
+layout: post
+
+title: Make easy loop on array
+tip-number: xx
+tip-username: jamet-julien
+tip-username-profile: https://github.com/jamet-julien
+tip-tldr: Make a elegant loop on array
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: Create an easy loop using an array

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    - fr
+---
+We often need to build loop, all the conditions can be huge:
+
+```js
+
+var index = 0,
+    aList = ['A','B','C', 'D', 'E']
+
+function change( dir){
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: You always have a space after `(`  ... either use it on both sides or remove it.

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    if( index < 0){
+      index = aList.length-1;
+    }
+
+    if( index >= aList.length ){
+      index = 0;
+    }
+
```
Comment (zenopopovici) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: Provide explanation here:

**Comment by zenopopovici** [comment]
@jamet-julien There is a bit of work to do maybe the others can chime in ...

**Comment by jamet-julien** [comment]
@zenopopovici Thanks.
Is it good now?

**Comment by zenopopovici** [comment]
@jamet-julien Thanks Julien, we will be adding this under review soon. We still have some backlog of tips proposals. When it is under review, more people will chime in and suggest changes. We'll ping you when that happens.

**Comment by jamet-julien** [comment]
@zenopopovici Thanks.

**Comment by gromgit** [comment]
@jamet-julien A few general comments:
- What does "the condition list would be huge" in your first sentence mean?
- Since `index` is part of the looping state of `aList`, it really should be a _property_ of `aList` (i.e. `aList.index`) rather than a separate variable.
- There's a bug in your example. Your `change()` function does a _pre-increment_, so the first `change()` call actually returns `aList[1]`, i.e. `B`.
- I think a more level-appropriate tip for JSTips would be a _loop generator_ function that takes an array as input and "enhances" it with self-looping capability:

``` js
function make_looper( arr ) {

  // Internal index to track next item to return
  arr.loop_idx = ???;  // your start index would be determined by whether
                  // next() and prev() are post- or pre-incremental

  // Return next item in loop
  arr.next = function() {
    // ...
  }

  // Return previous item in loop
  arr.prev = function() {
    // ...
  }

}

// example of pre-incrementing loop
var aList = ['A','B','C', 'D', 'E'];
aList.prev();  // -> E
aList.next();  // -> A
aList.next();  // -> B
```
- [Optional, for extra JSTips :+1:] You've looped to `E` in the above example, then someone calls `aList.pop()` twice, so `aList == ['A','B','C']`. How do you properly handle the next `aList.next()` or `aList.prev()`? (Your original `change()` function doesn't handle this correctly either.)

**Comment by loverajoel** [comment]
@gromgit really nice feedback

**Comment by jamet-julien** [comment]
@gromgit thank. I look this and update the code.

**Comment by gromgit** [comment]
@jamet-julien Much better now, although the first sentence is still rather vague. I suggest something like this instead:

> Sometimes, we need to loop endlessly over an array of items, like a carousel of images or an audio playlist. Here's how to take an array and give it "looping powers":

**Comment by jamet-julien** [comment]
@gromgit better, isn't it ?

**Comment by gromgit** [comment]
@jamet-julien :+1:

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+tip-username-profile: https://github.com/jamet-julien
+tip-tldr: Create an easy loop using an array 
+
+categories:
+    - en
+---
+Sometimes, we need to loop endlessly over an array of items, like a carousel of images or an audio playlist. Here's how to take an array and give it "looping powers":
+
+```js
+var aList = ['A','B','C', 'D', 'E'];
```
Comment (loverajoel) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: you have inconsistent spaces in the array

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+aList.prev();// -> B
+aList.prev();// -> A
+aList.prev();// -> D
+```
+
+Using the ```%``` operator is prettier:
+
+```js
+
+var aList = ['A','B','C', 'D', 'E'];
```
Comment (loverajoel) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: same here

File: _posts/en/2016-03-11-make-easy-loop-on-array.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+aList.next();// -> A
+aList.pop() ;// -> E
+aList.prev();// -> D
+aList.prev();// -> C
+aList.prev();// -> B
+aList.prev();// -> A
+aList.prev();// -> D
+```
+
+Using the ```%``` operator is prettier:
```
Comment (loverajoel) [inline _posts/en/2016-03-11-make-easy-loop-on-array.md]: can you add more explanation about the `%` use?

**Comment by jamet-julien** [comment]
@loverajoel thanks, is it Ok ?

**Comment by loverajoel** [comment]
@jamet-julien you have my :shipit:

**Comment by zenopopovici** [comment]
Mine too :shipit:

**Comment by jamet-julien** [comment]
@loverajoel  Can you tell me what is the next step please. Could I do modification ?
thank.

**Comment by loverajoel** [comment]
@JakeRawr we need 5 shipits to merge this, I'll try merge this week

**Comment by ppollo07** [comment]
:shipit: 🐔

**Comment by aschelstraete** [comment]
:shipit: 🎱

**Comment by sjfkai** [comment]
:shipit: 👍

**Comment by cjkumaresh** [comment]
:shipit:

**Comment by pklinger** [comment]
:shipit:

**Pull request merged**